File: C#/0501. Find Mode in Binary Search Tree.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public int[] FindMode(TreeNode root) {
        IList<int> result=new List<int>();
        if(root==null){
            return result.ToArray();
        }
        int cnt=0;
        int precnt=0;
        int target=int.MaxValue;
        Stack<TreeNode> rootStack=new Stack<TreeNode>();
        TreeNode node=root;
        while(node!=null || rootStack.Count()!=0){
            while(node!=null){
                rootStack.Push(node);
                node=node.left;
            }
            if(rootStack.Count!=0){
                TreeNode curNode=rootStack.Pop();
                if(target==int.MaxValue){
                    target=curNode.val;
                    cnt=1;
                }
                else if(target==curNode.val){
                    cnt+=1;
                }
                else{
                    if(cnt==precnt){
                        result.Add(target);
                        target=curNode.val;
                        cnt=1;
                    }
                    else if(cnt>precnt){
                        result.Clear();
                        result.Add(target);
                        precnt=cnt;
                        target=curNode.val;
                        cnt=1;
                    }
                    else{
                        target=curNode.val;
                        cnt=1;
                    }
                }
                node=curNode.right;
            }
        }
        if(cnt==precnt){
            result.Add(target);
        }
        if(cnt>precnt){
            result.Clear();
            result.Add(target);
        }
        return result.ToArray();
    }
}

File: archives/20190519python/0501. Find Mode in Binary Search Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def findMode(self, root: TreeNode) -> List[int]:
        rep=[]
        self.tree_to_list(root,rep)
        
        left=0
        cur_cnt=0
        result=[]
        while left<len(rep):
            cnt=1
            right=left+1
            while right<len(rep) and rep[right]==rep[left]:
                right+=1
                cnt+=1
            if cnt>cur_cnt:
                result=[rep[left]]
                cur_cnt=cnt
            elif cnt==cur_cnt:
                result.append(rep[left])
                cur_cnt=cnt
            left=right
        return result
        
    def tree_to_list(self,root,rep):
        if not root:
            return        
        self.tree_to_list(root.left,rep)
        rep.append(root.val)
        self.tree_to_list(root.right,rep)
